app/assets/javascripts/code_listing/code_listing.ts

```diff
@@ -281,77 +351,172 @@ export class CodeListing {
             }
         });
 
-        return node;
+        return form;
     }
 
-    async createAnnotation(lineId: number, text: string, form: HTMLFormElement): Promise<void> {
-        const response = await fetch(`/submissions/${this.submissionId}/annotations`, {
-            method: "POST",
-            headers: { "Content-Type": "application/json" },
-            body: JSON.stringify({
-                annotation: {
-                    // eslint-disable-next-line @typescript-eslint/camelcase
-                    line_nr: lineId,
-                    // eslint-disable-next-line @typescript-eslint/camelcase
-                    annotation_text: text
-                }
-            })
-        });
-        if (response.ok) {
-            this.addUserAnnotation(await response.json());
-            form.remove();
-        } else {
-            const error = await response.json();
-            const errorList: HTMLUListElement = UserAnnotation.processErrorMessage(error.responseJSON);
+    private createNewAnnotationForm(line: number | null): HTMLFormElement {
+        const onSubmit = async (form: HTMLFormElement): Promise<void> => {
+            const inputField = form.querySelector<HTMLTextAreaElement>("textarea");
+            inputField.classList.remove("validation-error");
+
+            const annotationData: UserAnnotationFormData = {
+                "annotation_text": inputField.value,
+                "line_nr": (line === null ? null : line - 1)
+            };
+
+            try {
+                const annotation = await UserAnnotation.create(annotationData, this.submissionId,
+                    (a, cb) => this.createUpdateAnnotationForm(a, cb));
+                this.addAnnotation(annotation);
+                form.remove();
+            } catch (err) {
+                inputField.classList.add("validation-error");
+            }
+        };
+
+        return this.createAnnotationForm(`annotation-submission-new-${line || 0}`,
+            null, onSubmit, form => form.remove());
+    }
+
+    private createUpdateAnnotationForm(annotation: UserAnnotation,
+        callback: CallableFunction): HTMLFormElement {
+        const onSubmit = async (form: HTMLFormElement): Promise<void> => {
+            const inputField = form.querySelector<HTMLTextAreaElement>("textarea");
+
+            const annotationData: UserAnnotationFormData = {
+                "annotation_text": inputField.value,
+                "line_nr": (annotation.line === null ? null : annotation.line - 1)
+            };
+
+            try {
+                const updated = await annotation.update(annotationData);
+                this.updateAnnotation(annotation, updated);
+            } catch (err) {
+                inputField.classList.add("validation-error");
+            }
+        };
+
+        const formId = `annotation-submission-update-${annotation.id}`;
+        return this.createAnnotationForm(formId, annotation, onSubmit, () => callback());
```

Comment:
```suggestion
        return this.createAnnotationForm(formId, annotation, onSubmit, callback);
```
Isn't this the same?

Reply:
This is not the same as it changes the context of `this` in the callback.